File: packages/x402-audit-api/tests/fixtures/cache_vuln.cpp
#include <map>
// Range proof verification with a memoized result, keyed on the output's txid.
static std::map<uint256, bool> g_proof_verify_cache;

bool VerifyRangeProof(const CTxOut& out, const CTxIn& in) {
    if (g_proof_verify_cache.count(out.txid)) {
        return true;
    }
    bool ok = secp256k1_rangeproof_verify(ctx, &min, &max, &out.commitment, out.proof.data());
    g_proof_verify_cache.insert(out.txid, ok);
    return ok;
}

File: packages/x402-audit-api/tests/fixtures/cache_vuln.rs
use std::collections::HashMap;
struct Verifier { proof_cache: HashMap<[u8;32], bool> }

impl Verifier {
    fn verify_proof(&mut self, out: &Output) -> Result<()> {
        if self.proof_cache.contains_key(&out.id) {
            return Ok(());
        }
        let ok = range_proof_verify(&out.commitment, &out.proof)?;
        self.proof_cache.insert(out.id, ok);
        Ok(())
    }
}
